src/MediatR/Mediator.cs

```diff
@@ -226,6 +256,25 @@ private IEnumerable<TWrapper> GetNotificationHandlers<TWrapper>(object notificat
                 .ToList();
         }
 
+        private Task<TResponse> GetPipeline<TResponse>(object request, RequestHandlerDelegate<TResponse> invokeHandler)
+        {
+            var requestType = request.GetType();
+            var method = CreatePipelineMethod.MakeGenericMethod(requestType, typeof(TResponse));

```

Comment:
That's not accurate at all, as what NServiceBus does can very easily be adapted to using a different instance every time, neatly supplied via the input arguments. This way you remove the state machine stack trace overhead but you still have all the freedom to input new instances on every call.

It is exactly that problem what I have been trying to solve in a performant manner, and it's very much possible.

The problem here is that you want to create a dynamic call chain `next -> next -> next` of length N, only way to do that is either code gen or allocate func's.

Problem with func is that you have to close over your current handler instance, which in effect makes the full pipeline 'single use'. 

The other option you have is pulling this instance out of the func and let the func execute a method on its input argument. Say `stateMachine.DoNext(next);` which would then internally advance one step and do correct async exception bubbling etc. I've seen an example of this and it's not pretty, lot's of work to get it right and in the end you still have a polluted stack trace. 

Your third option could be to supply a `ref int index` (only possible with a delegate) and an array of pipeline steps, so each func can pick the step at that index, increment the index and run the step. However then you need to have a way to carry the ref index with you when you call the step (as it has to call next) and ref params can be finicky.

When you look at all that work you could conclude, let's just code gen these delegate body's with the index number burned into them (pseudo code): 

```cs
public Task<TResponse> DelegateOne(IPipelineBehavior[] behaviors, request, next){
    return behaviors[0].Handle(request, next);
}

public Task<TResponse> DelegateTwo(IPipelineBehavior[] behaviors, request, next){
    return behaviors[1].Handle(request, next);
}

etc
``` 

Just cache these individually, then cache pipelines based on step count and you are good to go.